sparse/tests/test_coo.py

```diff
@@ -11,52 +11,88 @@
 from sparse.utils import assert_eq, random_value_array
 
 
-@pytest.mark.parametrize('reduction,kwargs,eqkwargs', [
-    ('max', {}, {}),
-    ('sum', {}, {}),
-    ('sum', {'dtype': np.float16}, {'atol': 1e-2}),
-    ('prod', {}, {}),
-    ('min', {}, {}),
+@pytest.fixture(scope='module', params=['f8', 'f4', 'i8', 'i4'])
+def random_sparse(request):
+    dtype = request.param
+    if np.issubdtype(dtype, np.integer):
+        def data_rvs(n):
+            return np.random.randint(-10000, 10000, n)
+    else:
+        data_rvs = None
+    return sparse.random((20, 30, 40), density=.25, data_rvs=data_rvs).astype(dtype)
+
+
+@pytest.mark.parametrize('reduction, kwargs', [
+    ('sum', {}),
+    ('sum', {'dtype': np.float32}),
+    ('mean', {}),
+    ('mean', {'dtype': np.float32}),
+    ('prod', {}),
+    ('max', {}),
+    ('min', {}),
 ])
 @pytest.mark.parametrize('axis', [None, 0, 1, 2, (0, 2), -3, (1, -1)])
 @pytest.mark.parametrize('keepdims', [True, False])
-def test_reductions(reduction, axis, keepdims, kwargs, eqkwargs):
-    x = sparse.random((2, 3, 4), density=.25)
+def test_reductions(reduction, random_sparse, axis, keepdims, kwargs):
+    x = random_sparse
     y = x.todense()
     xx = getattr(x, reduction)(axis=axis, keepdims=keepdims, **kwargs)
     yy = getattr(y, reduction)(axis=axis, keepdims=keepdims, **kwargs)
-    assert_eq(xx, yy, **eqkwargs)
+    assert_eq(xx, yy)
 
 
-@pytest.mark.parametrize('reduction,kwargs,eqkwargs', [
-    ('any', {}, {}),
-    ('all', {}, {}),
+@pytest.mark.xfail(reason=('Setting output dtype=float16 produces results '
+                           'inconsistent with numpy'))
+@pytest.mark.filterwarnings('ignore:overflow')
+@pytest.mark.parametrize('reduction, kwargs', [
+    ('sum', {'dtype': np.float16}),
+    ('mean', {'dtype': np.float16}),
```

Comment:
Optional: The reason we had `eqkwargs` for `float16` was large floating-point inaccuracies with `float16`. We're basically using [`np.allclose`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.allclose.html) to verify equivalence, and the `kwargs` in `assert_eq` are passed along to that.

It might be nice to rewrite this at some point, but that's how it is now. There is a plan to re-write this, basically something that takes two callables and some shapes and compares the NumPy and sparse functions (see #163).

Anyway, back to the point: If the `rtol` and `atol` parameters are increased, I'm pretty sure this will pass as well.

Reply:
> Anyway, back to the point: If the rtol and atol parameters are increased, I'm pretty sure this will pass as well.

This may be true, but the errors have to be pretty large. With `atol=10, rtol=10`, you still get errors for `sum`. This is an algorithmic difference, not just floating point error.

Reply:
> With `atol=10, rtol=10`, you still get errors for `sum`.

In that case, feel free to leave this as-is for now (and if you want, raise an issue documenting this).